examples/gateway-reshard.rs

```diff
@@ -61,73 +52,50 @@ async fn gateway_runner(client: Arc<Client>, mut shards: Vec<Shard>) {
 
                 continue;
             }
-            None => break,
         };
 
-        tokio::spawn(event_handler(Arc::clone(&client), event));
-    }
-}
-
-async fn event_handler(client: Arc<Client>, event: Event) -> anyhow::Result<()> {
-    match event {
-        Event::MessageCreate(message) if message.content == "!ping" => {
-            client
-                .create_message(message.channel_id)
-                .content("Pong!")?
-                .await?;
-        }
-        _ => {}
+        tracing::debug!(?event, shard = ?shard.id(), "received event");
     }
-
-    Ok(())
 }
 
-// Instrument to differentiate between the logs produced here and
-// in `gateway_runner`.
+// Instrument to differentiate between the logs produced here and in `runner`.
 #[tracing::instrument(skip_all)]
 async fn reshard(
     client: &Client,
     config: Config,
     config_callback: impl Fn(ShardId, ConfigBuilder) -> Config,
-) -> anyhow::Result<Option<Vec<Shard>>> {
+) -> anyhow::Result<Vec<Shard>> {
+    // Reshard every eight hours. This is an arbitrary number.
     const RESHARD_DURATION: Duration = Duration::from_secs(60 * 60 * 8);
 
-    // Reshard every eight hours.
     time::sleep(RESHARD_DURATION).await;
 
     let mut shards = stream::create_recommended(client, config, config_callback)
         .await?
         .collect::<Vec<_>>();
 
-    let mut identified = vec![false; shards.len()];
-    // Don't deserialize any events (with `ShardEventStream`) as the already
-    // running shards will handle them (the events are duplicated).
+    // Before swapping the old and new list of shards, try to identify them.
+    // Don't try too hard, however, as large bots may never have all shards
+    // identified at the same time.
+    let mut identified = vec![0; shards.len()];
     let mut stream = ShardMessageStream::new(shards.iter_mut());
 
-    // Drive the new list of shards until they are all identified.
-    while !identified.iter().all(|&shard| shard) {
+    while identified.iter().sum::<usize>() < (identified.len() * 3) / 4 {
```

Comment:
We should also be able to employ some better heuristics here if we want to take it further. Since we already call the get gateway endpoint, we can calculate the estimated time until which the shards *should* be identified. We could combine a timeout on that with the 75% identified goal to avoid returning a bit too early. Not sure how much benefit that little bit of added complexity would bring though.